spacy/pipeline/_parser_internals/_state.pxd

```diff
@@ -41,25 +40,29 @@ cdef cppclass StateC:
     int offset
     int _b_i
 
-    __init__(const TokenC* sent, int length) nogil:
+    __init__(const TokenC* sent, int length) nogil except +:
+        cdef int init_capacity
+
+        this._heads.resize(length, -1)
+        this._unshiftable.resize(length, False)
+
+        # Reserve memory ahead of time to minimize allocations during parsing.
+        # The initial capacity set here ideally reflects the expected average-case/majority usage.
+        init_capacity = 32
```

Comment:
I'd just use

```
    cdef int init_capacity = 32
```

here. Luckily this is not C89 :).

Reply:
Ah, I see. I wasn't sure that it was allowed, so I just followed the style I saw in the codebase. Will change these assignments.